File: app/src/main/java/com/example/a4g_ksd/MatchingAlgorithmTester.java
package com.example.a4g_ksd;

import java.util.ArrayList;

/**
 * Testing framework for the enhanced matching algorithm
 * 
 * @author Donateables Team
 *
 */
public class MatchingAlgorithmTester {
	
	/**
	 * Run comprehensive accuracy tests
	 */
	public static String runAccuracyTests() {
		StringBuilder results = new StringBuilder();
		results.append("Matching Algorithm Accuracy Tests\n");
		results.append("=====================================\n\n");
		
		// Test 1: Exact matches
		double exactMatchAccuracy = testExactMatches();
		results.append("Exact Matches: ").append(String.format("%.1f", exactMatchAccuracy)).append("%\n");
		
		// Test 2: Category matches
		double categoryMatchAccuracy = testCategoryMatches();
		results.append("Category Matches: ").append(String.format("%.1f", categoryMatchAccuracy)).append("%\n");
		
		// Test 3: Fuzzy matches
		double fuzzyMatchAccuracy = testFuzzyMatches();
		results.append("Fuzzy Matches: ").append(String.format("%.1f", fuzzyMatchAccuracy)).append("%\n");
		
		// Test 4: Real-world scenarios
		double realWorldAccuracy = testRealWorldScenarios();
		results.append("Real-world Scenarios: ").append(String.format("%.1f", realWorldAccuracy)).append("%\n");
		
		double overallAccuracy = (exactMatchAccuracy + categoryMatchAccuracy + fuzzyMatchAccuracy + realWorldAccuracy) / 4.0;
		results.append("\nOverall Accuracy: ").append(String.format("%.1f", overallAccuracy)).append("%\n");
		
		return results.toString();
	}
	
	/**
	 * Test exact name matching
	 */
	private static double testExactMatches() {
		ArrayList<Item> userItems = new ArrayList<>();
		userItems.add(new Item("Books", 5, "User"));
		userItems.add(new Item("Clothes", 3, "User"));
		userItems.add(new Item("Toys", 2, "User"));
		
		ArrayList<Item> orgItems = new ArrayList<>();
		orgItems.add(new Item("Books", 10, "Org"));
		orgItems.add(new Item("Clothes", 5, "Org"));
		orgItems.add(new Item("Toys", 8, "Org"));
		
		MatchingAlgorithm algorithm = new MatchingAlgorithm(userItems, orgItems);
		algorithm.matchItems();
		
		return algorithm.getAccuracy();
	}
	
	/**
	 * Test category-based matching
	 */
	private static double testCategoryMatches() {
		ArrayList<Item> userItems = new ArrayList<>();
		userItems.add(new Item("Textbooks", 5, "User"));
		userItems.add(new Item("Shirts", 3, "User"));
		userItems.add(new Item("Puzzles", 2, "User"));
		
		ArrayList<Item> orgItems = new ArrayList<>();
		orgItems.add(new Item("Books", 10, "Org"));
		orgItems.add(new Item("Clothing", 5, "Org"));
		orgItems.add(new Item("Toys", 8, "Org"));
		
		MatchingAlgorithm algorithm = new MatchingAlgorithm(userItems, orgItems);
		algorithm.matchItems();
		
		return algorithm.getAccuracy();
	}
	
	/**
	 * Test fuzzy string matching
	 */
	private static double testFuzzyMatches() {
		ArrayList<Item> userItems = new ArrayList<>();
		userItems.add(new Item("Book", 5, "User"));
		userItems.add(new Item("Cloth", 3, "User"));
		userItems.add(new Item("Toy", 2, "User"));
		
		ArrayList<Item> orgItems = new ArrayList<>();
		orgItems.add(new Item("Books", 10, "Org"));
		orgItems.add(new Item("Clothes", 5, "Org"));
		orgItems.add(new Item("Toys", 8, "Org"));
		
		MatchingAlgorithm algorithm = new MatchingAlgorithm(userItems, orgItems);
		algorithm.matchItems();
		
		return algorithm.getAccuracy();
	}
	
	/**
	 * Test real-world scenarios
	 */
	private static double testRealWorldScenarios() {
		ArrayList<Item> userItems = new ArrayList<>();
		userItems.add(new Item("Children's books", 5, "User"));
		userItems.add(new Item("Winter jackets", 3, "User"));
		userItems.add(new Item("Board games", 2, "User"));
		userItems.add(new Item("School supplies", 10, "User"));
		
		ArrayList<Item> orgItems = new ArrayList<>();
		orgItems.add(new Item("Books for kids", 10, "Org"));
		orgItems.add(new Item("Warm clothing", 5, "Org"));
		orgItems.add(new Item("Educational toys", 8, "Org"));
		orgItems.add(new Item("Pencils and paper", 20, "Org"));
		
		MatchingAlgorithm algorithm = new MatchingAlgorithm(userItems, orgItems);
		algorithm.matchItems();
		
		return algorithm.getAccuracy();
	}
}

File: iOS/Donateables/Donateables/MatchingAlgorithmTester.swift
import Foundation

// MARK: - Testing Framework
class MatchingAlgorithmTester {
    
    static func runAccuracyTests() -> String {
        var results = "Matching Algorithm Accuracy Tests\n"
        results += "=====================================\n\n"
        
        // Test 1: Exact matches
        let exactMatchAccuracy = testExactMatches()
        results += "Exact Matches: \(String(format: "%.1f", exactMatchAccuracy))%\n"
        
        // Test 2: Category matches
        let categoryMatchAccuracy = testCategoryMatches()
        results += "Category Matches: \(String(format: "%.1f", categoryMatchAccuracy))%\n"
        
        // Test 3: Fuzzy matches
        let fuzzyMatchAccuracy = testFuzzyMatches()
        results += "Fuzzy Matches: \(String(format: "%.1f", fuzzyMatchAccuracy))%\n"
        
        // Test 4: Real-world scenarios
        let realWorldAccuracy = testRealWorldScenarios()
        results += "Real-world Scenarios: \(String(format: "%.1f", realWorldAccuracy))%\n"
        
        let overallAccuracy = (exactMatchAccuracy + categoryMatchAccuracy + fuzzyMatchAccuracy + realWorldAccuracy) / 4.0
        results += "\nOverall Accuracy: \(String(format: "%.1f", overallAccuracy))%\n"
        
        return results
    }
    
    private static func testExactMatches() -> Double {
        let userItems = [
            Item(name: "Books", count: 5, origin: "User"),
            Item(name: "Clothes", count: 3, origin: "User"),
            Item(name: "Toys", count: 2, origin: "User")
        ]
        
        let orgItems = [
            Item(name: "Books", count: 10, origin: "Org"),
            Item(name: "Clothes", count: 5, origin: "Org"),
            Item(name: "Toys", count: 8, origin: "Org")
        ]
        
        let algorithm = MatchingAlgorithm(userItems: userItems, organizationItems: orgItems)
        algorithm.performMatching()
        
        return algorithm.getAccuracy()
    }
    
    private static func testCategoryMatches() -> Double {
        let userItems = [
            Item(name: "Textbooks", count: 5, origin: "User"),
            Item(name: "Shirts", count: 3, origin: "User"),
            Item(name: "Puzzles", count: 2, origin: "User")
        ]
        
        let orgItems = [
            Item(name: "Books", count: 10, origin: "Org"),
            Item(name: "Clothing", count: 5, origin: "Org"),
            Item(name: "Toys", count: 8, origin: "Org")
        ]
        
        let algorithm = MatchingAlgorithm(userItems: userItems, organizationItems: orgItems)
        algorithm.performMatching()
        
        return algorithm.getAccuracy()
    }
    
    private static func testFuzzyMatches() -> Double {
        let userItems = [
            Item(name: "Book", count: 5, origin: "User"),
            Item(name: "Cloth", count: 3, origin: "User"),
            Item(name: "Toy", count: 2, origin: "User")
        ]
        
        let orgItems = [
            Item(name: "Books", count: 10, origin: "Org"),
            Item(name: "Clothes", count: 5, origin: "Org"),
            Item(name: "Toys", count: 8, origin: "Org")
        ]
        
        let algorithm = MatchingAlgorithm(userItems: userItems, organizationItems: orgItems)
        algorithm.performMatching()
        
        return algorithm.getAccuracy()
    }
    
    private static func testRealWorldScenarios() -> Double {
        let userItems = [
            Item(name: "Children's books", count: 5, origin: "User"),
            Item(name: "Winter jackets", count: 3, origin: "User"),
            Item(name: "Board games", count: 2, origin: "User"),
            Item(name: "School supplies", count: 10, origin: "User")
        ]
        
        let orgItems = [
            Item(name: "Books for kids", count: 10, origin: "Org"),
            Item(name: "Warm clothing", count: 5, origin: "Org"),
            Item(name: "Educational toys", count: 8, origin: "Org"),
            Item(name: "Pencils and paper", count: 20, origin: "Org")
        ]
        
        let algorithm = MatchingAlgorithm(userItems: userItems, organizationItems: orgItems)
        algorithm.performMatching()
        
        return algorithm.getAccuracy()
    }
}